resources/assets/lib/interfaces/beatmapset-event-json.ts

```diff
@@ -4,23 +4,260 @@
 import BeatmapsetJson from './beatmapset-json';
 import GameMode from './game-mode';
 
-export function isBeatmapsetNominationEvent(x: BeatmapsetEventJson): x is BeatmapsetNominationEvent {
-  return x.type === 'nominate' && Array.isArray(x.comment?.modes);
+interface VoteEventJson {
+  score: number;
+  user_id: number;
 }
 
-export interface BeatmapsetNominationEvent extends BeatmapsetEventJson {
-  comment: {
-    modes: GameMode[];
-  };
-  type: 'nominate';
-}
-
-export default interface BeatmapsetEventJson {
+interface BaseBeatmapsetEvent {
   beatmapset?: BeatmapsetJson;
-  comment: any; // TODO: make always an object instead of object or string.
   created_at: string;
   discussion?: BeatmapsetDiscussionJson;
   id: number;
-  type: string;
   user_id?: number;
 }
+
+interface NominateEvent extends BaseBeatmapsetEvent {
+  comment: {
+    modes: GameMode[];
+  } | null;
+  type: 'nominate';
+}
+
+interface LoveEvent extends BaseBeatmapsetEvent {
+  comment: null;
+  type: 'love';
+}
+
+interface RemoveFromLovedEvent extends BaseBeatmapsetEvent {
+  comment: {
+    beatmap_discussion_id?: number;
+    beatmap_discussion_post_id?: number;
+    reason: string;
+  };
+  type: 'remove_from_loved';
+}
+
+interface QualifyEvent extends BaseBeatmapsetEvent {
+  comment: null;
+  type: 'qualify';
+}
+
+interface DisqualifyEvent extends BaseBeatmapsetEvent {
+  comment: {
+    beatmap_discussion_id?: number;
+    beatmap_discussion_post_id?: number;
+    nominator_ids?: number[];
+  } | string;
+  type: 'disqualify';
+}
+
+interface ApproveEvent extends BaseBeatmapsetEvent {
+  comment: null;
+  type: 'approve';
+}
+
+interface RankEvent extends BaseBeatmapsetEvent {
+  comment: null;
+  type: 'rank';
+}
+
+interface KudosuAllowEvent extends BaseBeatmapsetEvent {
+  comment: {
+    beatmap_discussion_id: number;
+    beatmap_discussion_post_id: number;
+  };
+  type: 'kudosu_allow';
+}
+
+interface KudosuDenyEvent extends BaseBeatmapsetEvent {
+  comment: {
+    beatmap_discussion_id: number;
+    beatmap_discussion_post_id: number;
+  };
+  type: 'kudosu_deny';
+}
+
+interface KudosuGainEvent extends BaseBeatmapsetEvent {
+  comment: {
+    beatmap_discussion_id?: number;
+    beatmap_discussion_post_id?: number;
+    new_vote: VoteEventJson;
+    votes: VoteEventJson[];
+  };
+  type: 'kudosu_gain';
+}
+
+interface KudosuLostEvent extends BaseBeatmapsetEvent {
+  comment: {
+    beatmap_discussion_id?: number;
+    beatmap_discussion_post_id?: number;
+    new_vote: VoteEventJson;
+    votes: VoteEventJson[];
+  };
+  type: 'kudosu_lost';
+}
+
+interface KudosuRecalculateEvent extends BaseBeatmapsetEvent {
+  comment: {
+    beatmap_discussion_id?: number;
+    beatmap_discussion_post_id?: number;
+  };
+  type: 'kudosu_recalculate';
+}
+
+interface IssueResolveEvent extends BaseBeatmapsetEvent {
+  comment: {
+    beatmap_discussion_id?: number;
```

Comment:
discussion_id for the issue-related events aren't nullable (but post_id is for oldest events)